sympy/core/expr.py

```diff
@@ -1160,30 +1183,31 @@ def as_terms(self):
         gens, terms = set(), []
 
         for term in Add.make_args(self):
-            coeff, _term = term.as_coeff_Mul()
-
-            coeff = complex(coeff)
-            cpart, ncpart = {}, []
-
-            if _term is not S.One:
-                for factor in Mul.make_args(_term):
-                    if factor.is_number:
-                        try:
-                            coeff *= complex(factor)
-                        except (TypeError, ValueError):
-                            pass
-                        else:
-                            continue
-
-                    if factor.is_commutative:
-                        base, exp = decompose_power(factor)
 
-                        cpart[base] = exp
-                        gens.add(base)
+            fac, ncpart = term.args_cnc()
+
+            # coeff is not tracking coeff, it is just
+            # tracking whether any numeric factor in a term has I
+            # in it and whether the number is more complicated
+            # than pure_complex form XXX
+            coeff = hasI, messy = [[1, (0,0)], 0]
+            cpart = defaultdict(lambda: S.Zero)
+            for factor in fac:
+                if factor.is_number:
```

Comment:
Look into whether numbers even have to be considered -- who uses `as_terms` and how? If only `as_ordered_terms` then this is probably not needed since touch up is done there.